src/Symfony/Component/Config/Resource/ReflectionClassResource.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Resource;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ReflectionClassResource implements SelfCheckingResourceInterface, \Serializable
+{
+    private $files = array();
+    private $className;
+    private $classReflector;
+    private $hash;
+
+    private static $mtimeCache = array();
+
+    public function __construct(\ReflectionClass $classReflector)
+    {
+        $this->className = $classReflector->name;
+        $this->classReflector = $classReflector;
+    }
+
+    public function isFresh($timestamp)
+    {
+        if (null === $this->hash) {
+            $this->hash = $this->computeHash();
+            $this->loadFiles($this->classReflector);
+        }
+
+        foreach ($this->files as $file => $v) {
+            if (!file_exists($file)) {
+                return false;
+            }
+
+            if (null === $mtime = &self::$mtimeCache[$file]) {
+                $mtime = @filemtime($file);
+            }
+            if ($mtime > $timestamp) {
+                if ($this->hash !== $this->computeHash()) {
+                    return false;
+                }
+
+                // ensure the next checks won't require a hash computation again,
+                // use a local mtime cache so that the current process still sees
+                // the original mtime and validates the hash for child classes
+                @touch($file, $timestamp);
```

Comment:
this might break other systems relying on the real mtime of the file because they are concerned about more things than the ones involved in the Symfony hash